pkg/tracex/trace.go

```diff
@@ -0,0 +1,241 @@
+// Package tracex implements a handshake tracer that can be passed to the TUN constructor to
+// observe handshake events.
+package tracex
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ooni/minivpn/internal/model"
+	"github.com/ooni/minivpn/internal/optional"
+	"github.com/ooni/minivpn/internal/session"
+)
+
+const (
+	handshakeEventStateChange = iota
+	handshakeEventPacketIn
+	handshakeEventPacketOut
+	handshakeEventPacketDropped
+)
+
+// HandshakeEventType indicates which event we logged.
+type HandshakeEventType int
+
+// Ensure that it implements the Stringer interface.
+var _ fmt.Stringer = HandshakeEventType(0)
+
+// String implements fmt.Stringer
+func (e HandshakeEventType) String() string {
+	switch e {
+	case handshakeEventStateChange:
+		return "state"
+	case handshakeEventPacketIn:
+		return "packet_in"
+	case handshakeEventPacketOut:
+		return "packet_out"
+	case handshakeEventPacketDropped:
+		return "packet_dropped"
+	default:
+		return "unknown"
+	}
+}
+
+// event is one handshake event collected by this [model.HandshakeTracer].
+type event struct {
+	// EventType is the type for this event.
+	EventType HandshakeEventType
+
+	// Stage is the stage of the handshake negotiation we're in.
+	Stage session.SessionNegotiationState
```

Comment:
this is the reason for the ints above. for this tracer implementation here it does not really matter since we can import. I was thinking that, not very far in the future, we'll have to expose at least a part of model (i.e., at the very minimal, options/config) and tun as public packages.

i see the point about API guarantees, but still I think it's more hygienic to hide some stuff as internal

I think I will go with option 1 for now, but knowing that we'll have to revisit public API real soon, and the tracer is probably the first case when this will be evident.

Reply:
Yeah, I was asking because I realized this is why you were hiding using `int`.

I think it's fine to change APIs if we need, especially if we avoid the hassle of v1 versions with Go.